src/representer.nim

```diff
@@ -0,0 +1,143 @@
+
+  if default.kind notin nnkLiterals:
+    default = default.normalizeValue(map)
+  # TODO: generic parameters
+  newIdentDefs(name, defType, default)
+
+proc normalizeRoutineDef(routineDef: NimNode, map: var IdentMap): NimNode =
+  ## RoutineDef Tree:
+  ##   Ident | Postfix(\*, Ident) (proc name)
+  ##   Empty # Related to Term rewriting macros which are not supported
+  ##   Empty | GenericParams
+  ##     IdentDefs
+  ##       Ident
+  ##       Type
+  ##   FormalParams
+  ##     empty | returntype
+  ##     @[IdentDefs] =
+  ##       idents | @[ident, ident]
+  ##       ident -> type
+  ##       empty | default
+  ##   Pragmas
+  ##   Empty
+  ##   StmtList # meat and potatoes
+  let formalParams = routineDef[3]
+  let returnType = formalParams[0]
+  let identDefs = formalParams[1..^1].mapIt(it.normalizeIdentDef(map))
+  routineDef.kind.newTree(
+    routineDef[0].addNewName(map), # RoutineDef Name
+    newEmptyNode(), # Term Rewriting macros and templates which are not supported
+    newEmptyNode(), # TODO: Implement Generic Params
+    nnkFormalParams.newTree(
+      returnType.normalizeValue(map)
+    ).add(identDefs),
+    newEmptyNode(), # TODO: Implement pragma
+    newEmptyNode(), # Empty, open for future use
+    routineDef.last.normalizeStmtList(map)
+  )
+
+proc normalizeImportExport(importStmt: NimNode, map: IdentMap): NimNode =
+  case importStmt.kind:
+  of nnkImportExceptStmt, nnkFromStmt, nnkExportExceptStmt:
+    importStmt.kind.newTree(importStmt[0]).add importStmt[1..^1].sortedByIt(it.strVal)
+  of nnkImportStmt, nnkExportStmt:
+    importStmt.kind.newTree(importStmt[0..^1].sortedByIt(if it.kind == nnkInfix: it.unpackInfix.left.strVal else: it.strVal)) # TODO: implemement normalizations of `import macros as m`
+  else:
+    raise newException(ValueError, $importStmt & "is not a valid import or export stmt")
+
+proc normalizeStmtList*(code: NimNode, map: var IdentMap): NimNode =
+  code.expectKind nnkStmtList
+  var normalizedTree = nnkStmtList.newTree
+
+  for index, statement in code:
+    normalizedTree.add case statement.kind:
+    of nnkCommentStmt: continue
+    of nnkVarSection..nnkConstSection:
+      var newDefSection = statement.kind.newTree()
+      for def in statement:
+        newDefSection.add def.normalizeIdentDef(map)
+      newDefSection
+    of nnkTypeSection: # TODO: Implement type normalizaation
+      statement
+    of RoutineNodes - {nnkLambda, nnkDo}: # We aren't supporting `do:` as of yet
+      statement.normalizeRoutineDef(map)
+    of nnkCallKinds:
+      statement.normalizeCall(map)
+    of nnkImportStmt, nnkFromStmt, nnkImportExceptStmt:
+      statement.normalizeImportExport(map)
+    of nnkDiscardStmt:
+      nnkDiscardStmt.newNimNode.add statement[0].normalizeValue(map)
+    else:
+      statement
+
+  normalizedTree
+
+proc createRepresentation*(fileName: string): (NimNode, IdentMap) =
+  var map: IdentMap
+  let code = parseStmt fileName.staticRead
+  (code.normalizeStmtList(map), map)
```

Comment:
Is there a specific reason that this proc should have the `result` variable? It isn't a very long proc?

Reply:
Mostly my personal preference - no big deal.

But I do think it's more idiomatic/readable to include `result =` when the `proc` isn't a one-liner or a single e.g. `if expression` or `case expression`. The Nim style guide doesn't take a position on it, but I think it's what I'm more used to reading.

Including `result =` makes it more immediately obvious that the purpose of a line is to change the return value of the proc, rather than e.g.:
- to modify a `var` parameter
- to call something that has a side effect.

That argument applies somewhat better to cases like in `normalizeIdentDef` above; the effect is smaller here, but I think it helps clarify the use of the anonymous field syntax for tuples as the return type.